stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+  std::vector<T_return_type> cc(n, 0);
+
+  for (size_t i = 0; i != n; i++) {
+    for (size_t j = 0; j != n; j++)
+      p2[j] = deltac[j] * g.p[0][i][j];
+
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p2[j];
+    T_return_type f2i = integrand(cc, pars);
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] - p2[j];
+    T_return_type temp = integrand(cc, pars);
+    f2i += temp;
+
+    for (size_t j = 0; j != n; j++)
+      p3[j] = deltac[j] * g.p[1][i][j];
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p3[j];
+    T_return_type f3i = integrand(cc, pars);
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] - p3[j];
+    temp = integrand(cc, pars);
+    f3i += temp;
+    f2 += f2i;
+    f3 += f3i;
+    divdiff[i] = fabs(f3i + twelvef1 - 7 * f2i);
+  }
+  free(p2);
+  free(p3);
+  T_return_type* p4
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  T_return_type f4 = 0.0;
+  for (size_t i = 0; i != g.p[2].size(); i++) {
+    for (size_t j = 0; j != n; j++)
+      p4[j] = deltac[j] * g.p[2][i][j];
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p4[j];
+    T_return_type temp = integrand(cc, pars);
+    f4 += temp;
+  }
+  free(p4);
+  T_return_type f5 = 0.0;
+  T_return_type* p5
+      = reinterpret_cast<T_return_type*>(malloc(n * sizeof(T_return_type)));
+  for (size_t i = 0; i != g.p[3].size(); i++) {
+    for (size_t j = 0; j != n; j++)
+      p5[j] = deltac[j] * g.p[3][i][j];
+
+    for (size_t j = 0; j != n; j++)
+      cc[j] = c[j] + p5[j];
+    T_return_type temp = integrand(cc, pars);
+    f5 += temp;
+  }
+  free(p5);
+  std::vector<T_return_type>().swap(cc);
+  T_return_type I
+      = v
+        * (g.w[0] * f1 + g.w[1] * f2 + g.w[2] * f3 + g.w[3] * f4 + g.w[4] * f5);
+  T_return_type Idash
+      = v * (g.wd[0] * f1 + g.wd[1] * f2 + g.wd[2] * f3 + g.wd[3] * f4);
+  T_return_type E = fabs(I - Idash);
+
+  int kdivide = 0;
+  T_return_type deltaf = E / (std::pow(10, n) * v);
+  for (size_t i = 0; i != n; i++) {
+    T_return_type delta = divdiff[i] - maxdivdiff;
+    if (delta > deltaf) {
+      kdivide = i;
+      maxdivdiff = divdiff[i];
+    } else if ((fabs(delta) <= deltaf) && (deltac[i] > deltac[kdivide])) {
+      kdivide = i;
+    }
+  }
+  out.result = I;
+  out.err = E;
+  out.kdivide = kdivide;
+  free(deltac);
+  free(divdiff);
+  std::vector<T_return_type>().swap(c);
+}
+
+class Box {
+ public:
+  Box(std::vector<double> a, std::vector<double> b, double& I, double& err,
+      int& kdivide)
+      : a(a), b(b), I(I), E(err), kdiv(kdivide) {}
+  bool operator<(const Box& box) const { return E > box.E; }
+  std::vector<double> a;
+  std::vector<double> b;
+  double I;
+  double E;
+  int kdiv;
+};
+
+inline class Box make_box(int n, std::vector<double> a, std::vector<double> b,
+                          one_d out) {
+  std::vector<double> ac(a);
+  std::vector<double> bc(b);
+  Box box(ac, bc, out.result, out.err, out.kdivide);
+  return box;
+}
```

Comment:
Is there a reason to not call the constructor directly?

Reply:
You mean to write instead of ms.insert(...make_box...) write:
  internal::Box box(a, b, out.result, out.err, out.kdivide);
  ms.insert(box);
on 3 spots and delete the make_box function?

Reply:
Comes with next commit.